src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java

```diff
@@ -7,297 +7,409 @@
+            Map<String, Double> participationPerBus = participatingElements.stream().collect(Collectors.toMap(
+                element -> getParticipatingElementBus(element).getId(),
+                element -> -element.getFactor(),
+                Double::sum
+            ));
+            getParticipationForBus = busId -> participationPerBus;
+        } else {
+            getParticipationForBus = busId -> Collections.singletonMap(lfNetwork.getSlackBus().getId(), -1d);
         }
+        computeFactorsInjection(getParticipationForBus, factorsGroups, new HashMap<>());
 
-        // initialize right hand side
-        DenseMatrix rhs = initRhs(lfNetwork, equationSystem, factorsByVarConfig);
+        Map<String, ComputedContingencyElement> contingenciesElements = contingencies.stream()
+                                                                                    .flatMap(contingency -> contingency.getElements().stream())
+                                                                                    .map(ComputedContingencyElement::new)
+                                                                                    .collect(Collectors.toMap(
+                                                                                        computedContingencyElement -> computedContingencyElement.getElement().getId(),
+                                                                                        computedContingencyElement -> computedContingencyElement
+                                                                                    ));
+
+        ComputedContingencyElement.setGlobalIndexes(contingenciesElements.values(), factorsGroups.size());
 
         // create jacobian matrix either using base network calculated voltages or nominal voltages
         VoltageInitializer voltageInitializer = lfParameters.getVoltageInitMode() == LoadFlowParameters.VoltageInitMode.PREVIOUS_VALUES ? new PreviousValueVoltageInitializer()
                 : new UniformValueVoltageInitializer();
         JacobianMatrix j = createJacobianMatrix(equationSystem, voltageInitializer);
+        // compute pre-contingency sensitivity values
+        DenseMatrix rhs = initRhs(lfNetwork, equationSystem, factorsGroups, contingenciesElements);
+        DenseMatrix states = solveTransposed(rhs, j); // states contains angles for the sensitivity factors, but also for the +1-1 related to contingencies
+
+        // sensitivities without contingency
+        List<SensitivityValue> sensitivityValues = calculateSensitivityValues(lfNetwork, equationSystem, factorsGroups,
+                states, functionReferenceByBranch, Collections.emptyList(), Collections.emptyMap());
 
-        // solve system
-        DenseMatrix states = solveTransposed(rhs, j);
+        Collection<Contingency> straightforwardContingencies = new LinkedList<>();
+        Map<List<Set<ComputedContingencyElement>>, List<Contingency>> losingConnectivityContingency = new HashMap<>();
```

Comment:
Not clear to me why you need this data structure, why do yu need to index by element that loose connectivity? It is very unlikely that 2 contingencies share the same elements.

Reply:
The idea was that if a given line is responsible for the loss of connectivity (let's say L1), then you don't have to recompute the participating factors for every N-2.
For example, all the contingencies [L1, L2], [L1, L3], [L1, Ln] will be indexed by the same `List<Set<ComputedContingencyElement>>`, and we will process all of them with only one solve.

Reply:
ok I understand